### ISSUE TITLE
posix.pipeline: please review

### ISSUE THREAD

**Issue by rrthomas** [luaposix/luaposix#74]
I've just added a function posix.pipeline to allow a pipeline of processes to be run simply from Lua, including both shell commands and Lua functions.

Please take a look and see what issues there are with the code. Obvious things to consider:
1. Should it handle some signals? Currently it doesn't, reasoning that Lua is not a shell, so if the signal ends up killing the calling process, that's fine; but that's not ideal for use at the REPL. (See e.g. http://www.cons.org/cracauer/sigint.html )
2. Should it die or return some error code on error? die-ing seems the logical thing to do to "return" from a process, but would it be nice to have some way to distinguish an error inside the pipeline function from errors in the called processes?
3. Is it really necessary to fork as much as it does? I think so, but I'm not entirely sure. In particular: the child shell processes can't use os.execute, as system(3) blocks SIGCHLD. It might seem unnecessary to fork again when the sub-process is actually a Lua function, but in fact it is, as otherwise it's not run in parallel, risking the pipeline blocking. But I'm not 100% convinced that it's necessary to fork AGAIN when there's another stage to the pipeline.
4. Any other hacks needed/not needed? For example, there are lots of calls to posix.close to make sure SIGPIPE will be delivered as per signal(7). The original stdout is saved and then restored at the end, but strictly that's only needed at the top level (I think?). Maybe the function could be restructured to avoid the nasty "if t2..." blocks.
5. What other features would be nice? The most obvious is a way to get the output of the last stage. Feeding input to the first stage is easily done by prepending an extra Lua function to the list, but owing to the one-way nature of pipes, and the fact that we're using fork here, it's harder to get data out: I propose to write a wrapper to set up a pipe and add an extra process to the end of the list to return the last stage's stdout like io.lines. Another idea would be to somehow wrap functions so they can take stdin as an argument and write to stdout with return values; this would require designing a calling convention that works with as little modification required to existing code as possible. For example, provide a pattern to split input into arguments, and specify a separator to string.join the results when writing them out.

I think this is a really interesting feature (I've been meaning to implement it for months) to make it easier to write shell scripts in Lua. I don't know of a similar feature in other scripting languages, either. In general I think it should work "just like" the shell, and extra functionality should be wrapped on top. But getting it right is tricky both in terms of robustness and exact API, so I'm very open to suggestions.
